File: experiments/src/actions/decisionactions.jl
module DecisionActionsModule

using ..ActionsModule
using ..DecisionNetModule

@kwdef struct DecisionActions <: AbstractActions
    meta_actions::Dict{Symbol, Vector{Symbol}}
    min_thresholds::Vector{Float64}
    max_thresholds::Vector{Float64}
    n_thresholds::Int
    allow_refs::Bool
    allow_cycles::Bool
end
export DecisionActions

function decision_actions_list(actions::DecisionActions)::Vector{Symbol}
    meta_keys = keys(actions.meta_actions)
    meta_keys = collect(meta_keys)

    list = vcat([:NEXT_FEATURE, :NEXT_THRESHOLD, :NEXT_NODE, :SELECT_NODE, :SET_TRUE_IDX, :SET_FALSE_IDX, :NEW_BRANCH_NODE], meta_keys)

    if actions.allow_refs
        push!(list, :NEW_REF_NODE)
    end

    return list
end
export decision_actions_list

function decision_action!(net::DecisionNet, action::Symbol, state::ActionsState, actions::DecisionActions)

    nodes = net.nodes
    meta_actions = actions.meta_actions

    node_idx = state.node_idx
    selected_idx = state.selected_idx

    if haskey(meta_actions, action)
        for sub_action = meta_actions[action]
            decision_action!(net, sub_action, state, actions)
        end
    elseif action == :NEXT_FEATURE
        n_features = length(actions.min_thresholds)
        
        state.feature_idx += 1
        if state.feature_idx > n_features
            state.feature_idx = 1
        end
    elseif action == :NEXT_THRESHOLD
        state.threshold_idx += 1

        if state.threshold_idx > actions.n_thresholds
            state.threshold_idx = 1
        end
    elseif action == :NEW_BRANCH_NODE
        threshold = threshold_value(state, actions)

        new_node = BranchNode(
            threshold = threshold,
            feature_idx = state.feature_idx
        )
        
        push!(nodes, new_node)
    elseif action == :NEW_REF_NODE
        new_node = RefNode(ref_idx = selected_idx)
        
        push!(nodes, new_node)
    elseif !isempty(nodes)

        node = nodes[node_idx]

        is_feedforward = selected_idx > node_idx

        if action == :NEXT_NODE
            state.node_idx += 1
            
            nodes_len = length(nodes)
            if state.node_idx > nodes_len
                state.node_idx = 1
            end
        elseif action == :SELECT_NODE
            state.selected_idx = node_idx
        elseif actions.allow_cycles || is_feedforward
            if action == :SET_TRUE_IDX
                node.true_idx = selected_idx
            elseif action == :SET_FALSE_IDX
                node.false_idx = selected_idx
            end
        end
    end
end
export decision_action!

end

File: experiments/alphchemy/src/actions/decision_actions.rs
use std::collections::HashMap;
use serde_json::{json, Value};

use crate::actions::actions::{Action, Actions, ActionsState, ThresholdRange, meta_actions_json, thresholds_json};
use crate::network::decision_net::{DecisionNet, DecisionNode, BranchNode, RefNode};

#[derive(Clone, Debug)]
pub struct DecisionActions {
    pub meta_actions: HashMap<String, Vec<Action>>,
    pub thresholds: HashMap<String, ThresholdRange>,
    pub feat_order: Vec<String>,
    pub n_thresholds: usize,
    pub allow_refs: bool
}

impl DecisionActions {
    pub fn to_json(&self) -> Value {
        json!({
            "type": "decision",
            "meta_actions": meta_actions_json(&self.meta_actions),
            "thresholds": thresholds_json(&self.thresholds, &self.feat_order),
            "feat_order": self.feat_order,
            "n_thresholds": self.n_thresholds,
            "allow_refs": self.allow_refs
        })
    }
}

impl Actions<DecisionNet> for DecisionActions {
    fn actions_list(&self) -> Vec<Action> {
        let mut list = vec![Action::NextFeat, Action::NextThreshold, Action::NextNode, Action::SelectNode, Action::SetFeat, Action::SetThreshold, Action::SetTrueIdx, Action::SetFalseIdx, Action::SetRefIdx, Action::NewBranch, Action::NewRef];

        for label in self.meta_actions.keys() {
            list.push(Action::MetaAction(label.clone()));
        }

        list
    }

    fn do_action(&self, net: &mut DecisionNet, state: &mut ActionsState, action: Action) {
        let node_idx = state.node_idx;
        let selected_idx = state.selected_idx;

        match action {
            Action::MetaAction(label) => {
                if let Some(sub_actions) = self.meta_actions.get(&label) {
                    for sub_action in sub_actions {
                        self.do_action(net, state, sub_action.clone());
                    }
                }
            }
            Action::NextFeat => state.next_feat(self.feat_order.len()),
            Action::NextThreshold => state.next_threshold(self.n_thresholds),
            Action::NextNode => state.next_node(net.nodes.len()),
            Action::SelectNode => state.select_node(),
            Action::SetFeat => {
                let maybe_feat_id = self.feat_order.get(state.feat_idx);

                if let Some(feat_id) = maybe_feat_id
                && let Some(node) = net.nodes.get_mut(node_idx)
                && let DecisionNode::Branch(branch_node) = node {

                    branch_node.feat_id = Some(feat_id.clone());
                }
            }
            Action::SetThreshold => {
                let maybe_node = net.nodes.get_mut(node_idx);

                if let Some(DecisionNode::Branch(branch_node)) = maybe_node
                && let Some(feat_id) = branch_node.feat_id.clone()
                && let Some(range) = self.thresholds.get(&feat_id) {

                    let threshold = range.value_at(state.threshold_idx, self.n_thresholds);
                    branch_node.threshold = Some(threshold);
                }
            }
            Action::SetTrueIdx => {
                if let Some(node) = net.nodes.get_mut(node_idx) {
                    node.set_true_idx(selected_idx);
                }
            }
            Action::SetFalseIdx => {
                if let Some(node) = net.nodes.get_mut(node_idx) {
                    node.set_false_idx(selected_idx);
                }
            }
            Action::SetRefIdx => {
                if let Some(node) = net.nodes.get_mut(node_idx)
                && let DecisionNode::Ref(ref_node) = node {

                    ref_node.ref_idx = Some(selected_idx);
                }
            }
            Action::NewBranch => {
                let branch_node = BranchNode {
                    threshold: None,
                    feat_id: None,
                    true_idx: None,
                    false_idx: None,
                    value: false
                };
                let new_node = DecisionNode::Branch(branch_node);
                net.nodes.push(new_node);
            }
            Action::NewRef => {
                if self.allow_refs {
                    let ref_node = RefNode {
                        ref_idx: None,
                        true_idx: None,
                        false_idx: None,
                        value: false
                    };
                    let new_node = DecisionNode::Ref(ref_node);
                    net.nodes.push(new_node);
                }
            }
            _ => ()
        }
    }
}
